Hago un alimpieza de datos, 
se habian analizado las columnas que no aportan nada. 
algunas columnas tienen una estructura un poco sucia para el entrenamiento
de  un solo paso converti esos jsons a columnas booleanas, y regreso una lista limpa.
por si encuentro un metodo nuev mejor para usaar estas columnas o valores en el entrenamiento.

In [ ]:
import pandas as pd
import glob
pd.set_option('display.max_columns', None)


# This code list all csvs in a folder, and merge them all, remove all the values that are not rated,

In [2]:
merged_df= pd.read_csv('_ALL_games_pc-console_2000-2023.csv')


In [3]:
print(merged_df.shape)
rated_df=merged_df[merged_df['rating']>0]
not_rated_df=merged_df[merged_df['rating']==0]
print("Not rated",not_rated_df.shape)
print("Rated",rated_df.shape)
rated_df.head()

(102545, 32)
Not rated (81308, 32)
Rated (21237, 32)


slug  \
0                           halo-infinite   
1                              elden-ring   
2                                 stray-2   
3  uncharted-legacy-of-thieves-collection   
4                         forza-horizon-5   

                                      name  playtime  \
0                            Halo Infinite         2   
1                               Elden Ring        59   
2                                    Stray         5   
3  Uncharted: Legacy of Thieves Collection         6   
4                          Forza Horizon 5        14   

                                           platforms  \
0  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
1  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
2  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
3  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   
4  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   

                                              stores    released    tba  \
0  [{'store': {'id': 1, 'name': 'Steam', 'slug': ...  2021-12-08  False   
1  [{'store': {'id': 1, 'name': 'Steam', 'slug': ...  2022-02-25  False   
2  [{'store': {'id': 1, 'name': 'Steam', 'slug': ...  2022-07-19  False   
3  [{'store': {'id': 1, 'name': 'Steam', 'slug': ...  2022-01-28  False   
4  [{'store': {'id': 1, 'name': 'Steam', 'slug': ...  2021-11-08  False   

                                    background_image  rating  rating_top  \
0  https://media.rawg.io/media/games/e1f/e1ffbeb1...    3.68           4   
1  https://media.rawg.io/media/games/5ec/5ecac5cb...    4.40           5   
2  https://media.rawg.io/media/games/cd3/cd3c9c7d...    4.16           4   
3  https://media.rawg.io/media/games/de6/de66bc4c...    4.41           5   
4  https://media.rawg.io/media/games/082/08236550...    4.32           5   

                                             ratings  ratings_count  \
0  [{'id': 4, 'title': 'recommended', 'count': 28...            619   
1  [{'id': 5, 'title': 'exceptional', 'count': 62...            917   
2  [{'id': 4, 'title': 'recommended', 'count': 35...            809   
3  [{'id': 5, 'title': 'exceptional', 'count': 19...            352   
4  [{'id': 5, 'title': 'exceptional', 'count': 29...            598   

   reviews_text_count  added  \
0                  10   6818   
1                  26   6111   
2                  12   4362   
3                   4   4246   
4                  10   4146   

                                     added_by_status  metacritic  \
0  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...        84.0   
1  {'yet': 312, 'owned': 3757, 'beaten': 664, 'to...        95.0   
2  {'yet': 276, 'owned': 2142, 'beaten': 842, 'to...        82.0   
3  {'yet': 197, 'owned': 3366, 'beaten': 313, 'to...        88.0   
4  {'yet': 128, 'owned': 2978, 'beaten': 229, 'to...        92.0   

   suggestions_count              updated      id  score  clip  \
0                883  2023-07-29T14:33:43   58751    NaN   NaN   
1                573  2023-07-29T12:11:54  326243    NaN   NaN   
2                128  2023-07-29T14:35:09  452638    NaN   NaN   
3                485  2023-07-29T07:07:04  704634    NaN   NaN   
4                520  2023-07-28T03:56:49  622492    NaN   NaN   

                                                tags  \
0  [{'id': 31, 'name': 'Singleplayer', 'slug': 's...   
1  [{'id': 31, 'name': 'Singleplayer', 'slug': 's...   
2  [{'id': 31, 'name': 'Singleplayer', 'slug': 's...   
3  [{'id': 31, 'name': 'Singleplayer', 'slug': 's...   
4  [{'id': 31, 'name': 'Singleplayer', 'slug': 's...   

                                         esrb_rating  user_game  \
0  {'id': 3, 'name': 'Teen', 'slug': 'teen', 'nam...        NaN   
1  {'id': 4, 'name': 'Mature', 'slug': 'mature', ...        NaN   
2  {'id': 2, 'name': 'Everyone 10+', 'slug': 'eve...        NaN   
3  {'id': 3, 'name': 'Teen', 'slug': 'teen', 'nam...        NaN   
4                                                NaN        NaN   

   reviews_count s

In [4]:

# not_rated_df

In [5]:
rated_df.columns

Index(['slug', 'name', 'playtime', 'platforms', 'stores', 'released', 'tba',
       'background_image', 'rating', 'rating_top', 'ratings', 'ratings_count',
       'reviews_text_count', 'added', 'added_by_status', 'metacritic',
       'suggestions_count', 'updated', 'id', 'score', 'clip', 'tags',
       'esrb_rating', 'user_game', 'reviews_count', 'saturated_color',
       'dominant_color', 'short_screenshots', 'parent_platforms', 'genres',
       'community_rating', 'Unnamed: 0'],
      dtype='object')

In [6]:
rated_df.iloc[0]

slug                                                      halo-infinite
name                                                      Halo Infinite
playtime                                                              2
platforms             [{'platform': {'id': 4, 'name': 'PC', 'slug': ...
stores                [{'store': {'id': 1, 'name': 'Steam', 'slug': ...
released                                                     2021-12-08
tba                                                               False
background_image      https://media.rawg.io/media/games/e1f/e1ffbeb1...
rating                                                             3.68
rating_top                                                            4
ratings               [{'id': 4, 'title': 'recommended', 'count': 28...
ratings_count                                                       619
reviews_text_count                                                   10
added                                                           

# COLUMNS NOT USEFUL <<<<<<<<<
'tags', 'unnamed: 0', 'community_rating','short_screenshots','clip','updated','dominant_color', 'saturated_color', 'background_image', 'tba','reviews_text_count','suggestions_count'
- tags: contienne mucha informacion que no se a que pertenece, como imagenes de otros juegos, comentario, o texto en ruso
- unnamed: 0 , a remnant index from the scrapping/fetching
- community_rating, a value that may be useful, but it is not present in most rows
- short_screenshots, urls for images
- clip, url for videos
- updated, date showing if the record has been updated
- colors?,[dominant_color, saturated_color] a list of colors, not sure based on what
- background_image, image used for the page
- tba: to be announced, not released yet
- reviews_text_count
- suggestions_count ???

### explode categories: platforms, stores, ratings, added_by_status, esrb_rating, parent_platforms, genres
### useless columns: unnamed: 0 , tags, comminity_rating, short_screenshots, clip, updated, colors?, background_image

parent_platforms: For instance, for PS2 and PS4 the “parent platform” is PlayStation.



# Convert column with JSONs value, and split it into different columns

In [8]:
import ast
import json

df=rated_df.copy()

# Function to convert the string representation to a list of dictionaries
# def convert_to_list(row):
#     return ast.literal_eval(row['platforms'])
def convert_to_list(column_data):
    if (column_data == "" or column_data=="nan"):
        return []    
    try:
        return ast.literal_eval(column_data)
    except (ValueError, SyntaxError):
        return []
    
def convert_json_to_list(column_data):
    try:
        return json.loads(column_data)
    except json.JSONDecodeError:
        return []
    
    
columns_to_drop =['tags', 'Unnamed: 0', 'community_rating','short_screenshots','clip','updated','dominant_color', 'saturated_color', 'background_image', 'tba','reviews_text_count','suggestions_count']
df=df.drop(columns_to_drop, axis=1)
df.head(1)

slug           name  playtime  \
0  halo-infinite  Halo Infinite         2   

                                           platforms  \
0  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...   

                                              stores    released  rating  \
0  [{'store': {'id': 1, 'name': 'Steam', 'slug': ...  2021-12-08    3.68   

   rating_top                                            ratings  \
0           4  [{'id': 4, 'title': 'recommended', 'count': 28...   

   ratings_count  added                                    added_by_status  \
0            619   6818  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...   

   metacritic     id  score  \
0        84.0  58751    NaN   

                                         esrb_rating  user_game  \
0  {'id': 3, 'name': 'Teen', 'slug': 'teen', 'nam...        NaN   

   reviews_count                                   parent_platforms  \
0            633  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   

                                              genres  
0  [{'id': 2, 'name': 'Shooter', 'slug': 'shooter...

## 1- This extract the platforms on each game is distributed
A column per category is created, and a new one containin a list of all.

In [9]:
rated_df['platforms'].iloc[0] #{'platform': {'id': 4, 'name': 'PC', 'slug':"xyz"

"[{'platform': {'id': 4, 'name': 'PC', 'slug': 'pc'}}, {'platform': {'id': 1, 'name': 'Xbox One', 'slug': 'xbox-one'}}, {'platform': {'id': 186, 'name': 'Xbox Series S/X', 'slug': 'xbox-series-x'}}]"

In [10]:
platforms_cols=[]
# Function to extract 'slug' values and create binary columns
def extract_platforms_slug_values(row):
    # parent_platforms, stores
    platforms = row['platforms']
    slugs = [platform['platform']['slug'] for platform in platforms]
    row['platforms_list'] = slugs

    for slug in slugs:
        row[f"platform_{slug}"] = True
        platforms_cols.append(row[f"platform_{slug}"])
    return row


# Make sure the type of the column is set to list and not string
# df['platforms'] = df.apply(convert_to_list, axis=1)
df['platforms'] = df['platforms'].apply(convert_to_list)

# Apply the function to each row of the DataFrame
df = df.apply(extract_platforms_slug_values, axis=1)

# Fill NaN values with False (for platforms not present in a particular row)
# df['platforms'].fillna(False, inplace=True)

platforms_cols=list(set(platforms_cols))
df[platforms_cols]=df[platforms_cols].fillna(False)

# Drop the original 'platforms' column (optional)
df.drop('platforms', axis=1, inplace=True)
df.rename(columns={'platforms_list':"platforms"}, inplace=True)
df.head(1)

added                                    added_by_status  \
0   6818  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...   

                                         esrb_rating  \
0  {'id': 3, 'name': 'Teen', 'slug': 'teen', 'nam...   

                                              genres     id  metacritic  \
0  [{'id': 2, 'name': 'Shooter', 'slug': 'shooter...  58751        84.0   

            name                                   parent_platforms  \
0  Halo Infinite  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   

  platform_android platform_apple-ii platform_commodore-amiga  \
0              NaN               NaN                      NaN   

  platform_dreamcast platform_game-boy platform_game-boy-advance  \
0                NaN               NaN                       NaN   

  platform_game-boy-color platform_gamecube platform_genesis platform_ios  \
0                     NaN               NaN              NaN          NaN   

  platform_jaguar platform_linux platform_macintosh platform_macos  \
0             NaN            NaN                NaN            NaN   

  platform_neogeo platform_nes platform_nintendo-3ds platform_nintendo-64  \
0             NaN          NaN                   NaN                  NaN   

  platform_nintendo-ds platform_nintendo-dsi platform_nintendo-switch  \
0                  NaN                   NaN                      NaN   

  platform_pc platform_playstation1 platform_playstation2  \
0        True                   NaN                   NaN   

  platform_playstation3 platform_playstation4 platform_playstation5  \
0                   NaN                   NaN                   NaN   

  platform_ps-vita platform_psp platform_sega-cd platform_sega-master-system  \
0              NaN          NaN              NaN                         NaN   

  platform_sega-saturn platform_web platform_wii platform_wii-u  \
0                  NaN          NaN          NaN            NaN   

  platform_xbox-old platform_xbox-one platform_xbox-series-x platform_xbox360  \
0               NaN              True                   True              NaN   

                       platforms  playtime  rating  rating_top  \
0  [pc, xbox-one, xbox-series-x]         2    3.68           4   

                                             ratings  ratings_count  \
0  [{'id': 4, 'title': 'recommended', 'count': 28...            619   

     released  reviews_count  score           slug  \
0  2021-12-08            633    NaN  halo-infinite   

                                              stores  user_game  
0  [{'store': {'id': 1, 'name': 'Steam', 'slug': ...        NaN

## 2- Extract the genres


In [11]:
rated_df['genres'].iloc[0]

"[{'id': 2, 'name': 'Shooter', 'slug': 'shooter'}, {'id': 3, 'name': 'Adventure', 'slug': 'adventure'}, {'id': 4, 'name': 'Action', 'slug': 'action'}]"

In [12]:
genres_cols=[]
def extract_genres(row):
    genres_list = row['genres']
    slugs = [genre['slug'] for genre in genres_list]
    row['genres']=slugs
    for slug in slugs:
        row[f"genre_{slug}"] = True
        genres_cols.append(row[f"genre_{slug}"])
    return row

# Apply the conversion function to the 'genres' column
df['genres'] = df['genres'].apply(convert_to_list)
df = df.apply(extract_genres, axis=1)

genres_cols=list(set(genres_cols))
df[genres_cols]=df[genres_cols].fillna(False)

df['genres'].fillna("", inplace=True)
df.head(2)

added                                    added_by_status  \
0   6818  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...   
1   6111  {'yet': 312, 'owned': 3757, 'beaten': 664, 'to...   

                                         esrb_rating genre_action  \
0  {'id': 3, 'name': 'Teen', 'slug': 'teen', 'nam...         True   
1  {'id': 4, 'name': 'Mature', 'slug': 'mature', ...         True   

  genre_adventure genre_arcade genre_board-games genre_card genre_casual  \
0            True          NaN               NaN        NaN          NaN   
1             NaN          NaN               NaN        NaN          NaN   

  genre_educational genre_family genre_fighting genre_indie  \
0               NaN          NaN            NaN         NaN   
1               NaN          NaN            NaN         NaN   

  genre_massively-multiplayer genre_platformer genre_puzzle genre_racing  \
0                         NaN              NaN          NaN          NaN   
1                         NaN              NaN          NaN          NaN   

  genre_role-playing-games-rpg genre_shooter genre_simulation genre_sports  \
0                          NaN          True              NaN          NaN   
1                         True           NaN              NaN          NaN   

  genre_strategy                            genres      id  metacritic  \
0            NaN      [shooter, adventure, action]   58751        84.0   
1            NaN  [action, role-playing-games-rpg]  326243        95.0   

            name                                   parent_platforms  \
0  Halo Infinite  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   
1     Elden Ring  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   

  platform_android platform_apple-ii platform_commodore-amiga  \
0              NaN               NaN                      NaN   
1              NaN               NaN                      NaN   

  platform_dreamcast platform_game-boy platform_game-boy-advance  \
0                NaN               NaN                       NaN   
1                NaN               NaN                       NaN   

  platform_game-boy-color platform_gamecube platform_genesis platform_ios  \
0                     NaN               NaN              NaN          NaN   
1                     NaN               NaN              NaN          NaN   

  platform_jaguar platform_linux platform_macintosh platform_macos  \
0             NaN            NaN                NaN            NaN   
1             NaN            NaN                NaN            NaN   

  platform_neogeo platform_nes platform_nintendo-3ds platform_nintendo-64  \
0             NaN          NaN                   NaN                  NaN   
1             NaN          NaN                   NaN                  NaN   

  platform_nintendo-ds platform_nintendo-dsi platform_nintendo-switch  \
0                  NaN                   NaN                      NaN   
1                  NaN                   NaN                      NaN   

  platform_pc platform_playstation1 platform_playstation2  \
0        True                   NaN                   NaN   
1        True                   NaN                   NaN   

  platform_playstation3 platform_playstation4 platform_playstation5  \
0                   NaN                   NaN                   NaN   
1                   NaN                  True                  True   

  platform_ps-vita platform_psp platform_sega-cd platform_sega-master-system  \
0              NaN          NaN              NaN                         NaN   
1              NaN          NaN              NaN                         NaN   

  platform_sega-saturn platform_web platform_wii platform_wii-u  \
0                  NaN          NaN          NaN            NaN   
1                  NaN          NaN          NaN            NaN   

  platform_xbox-old platform_xbox-one platform_xbox-series-x platform_xbox360  \
0               NaN              True                   True       

## 3- Extract the stores 
Where the games are being sold, and create a column for each one

In [13]:
print(rated_df['stores'].loc[231])
print(rated_df['stores'].loc[0])


nan
[{'store': {'id': 1, 'name': 'Steam', 'slug': 'steam'}}, {'store': {'id': 2, 'name': 'Xbox Store', 'slug': 'xbox-store'}}]


In [37]:
all_stores_columns=[]
def extract_stores(row):
    stores_list = row['stores']
    slugs = [store['store']['slug'] for store in stores_list]
    row['stores']=slugs
    for slug in slugs:
        row[f"store_{slug}"] = True
        all_stores_columns.append(f"store_{slug}")
    return row

df['stores'] = df['stores'].astype(str)
df['stores'] = df['stores'].apply(convert_to_list)

df = df.apply(extract_stores, axis=1)

all_stores_columns=list(set(all_stores_columns))
df[all_stores_columns]=df[all_stores_columns].fillna(False)

df['stores'].fillna("", inplace=True)
df.head(2)

Unnamed: 0  added                                    added_by_status  \
0         NaN   6818  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...   
1         NaN   6111  {'yet': 312, 'owned': 3757, 'beaten': 664, 'to...   

                                    background_image  clip  community_rating  \
0  https://media.rawg.io/media/games/e1f/e1ffbeb1...   NaN               NaN   
1  https://media.rawg.io/media/games/5ec/5ecac5cb...   NaN               NaN   

  dominant_color                                        esrb_rating  \
0         0f0f0f  {'id': 3, 'name': 'Teen', 'slug': 'teen', 'nam...   
1         0f0f0f  {'id': 4, 'name': 'Mature', 'slug': 'mature', ...   

                                              genres      id  metacritic  \
0  [{'id': 2, 'name': 'Shooter', 'slug': 'shooter...   58751        84.0   
1  [{'id': 4, 'name': 'Action', 'slug': 'action'}...  326243        95.0   

            name                                   parent_platforms  \
0  Halo Infinite  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   
1     Elden Ring  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   

                                           platforms  playtime  rating  \
0  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...         2    3.68   
1  [{'platform': {'id': 4, 'name': 'PC', 'slug': ...        59    4.40   

   rating_top                                            ratings  \
0           4  [{'id': 4, 'title': 'recommended', 'count': 28...   
1           5  [{'id': 5, 'title': 'exceptional', 'count': 62...   

   ratings_count    released  reviews_count  reviews_text_count  \
0            619  2021-12-08            633                  10   
1            917  2022-02-25            955                  26   

  saturated_color  score                                  short_screenshots  \
0          0f0f0f    NaN  [{'id': -1, 'image': 'https://media.rawg.io/me...   
1          0f0f0f    NaN  [{'id': -1, 'image': 'https://media.rawg.io/me...   

            slug store_apple-appstore store_epic-games store_gog  \
0  halo-infinite                  NaN              NaN       NaN   
1     elden-ring                  NaN              NaN       NaN   

  store_google-play store_itch store_nintendo store_playstation-store  \
0               NaN        NaN            NaN                     NaN   
1               NaN        NaN            NaN                    True   

  store_steam store_xbox-store store_xbox360  \
0        True             True           NaN   
1        True             True           NaN   

                                   stores  suggestions_count  \
0                     [steam, xbox-store]                883   
1  [steam, playstation-store, xbox-store]                573   

                                                tags    tba  \
0  [{'id': 31, 'name': 'Singleplayer', 'slug': 's...  False   
1  [{'id': 31, 'name': 'Singleplayer', 'slug': 's...  False   

               updated  user_game  
0  2023-07-29T14:33:43        NaN  
1  2023-07-29T12:11:54        NaN

## 4- Extract type of ratings
This code extract from ratings the ammount of type of rates
'rating_top', rating with ceil()
'ratings', type of ratings
'ratings_count', the sum of values in type of ratings: in general the total of ratings

In [15]:
rated_df['ratings'].iloc[0]

"[{'id': 4, 'title': 'recommended', 'count': 287, 'percent': 45.34}, {'id': 3, 'title': 'meh', 'count': 171, 'percent': 27.01}, {'id': 5, 'title': 'exceptional', 'count': 124, 'percent': 19.59}, {'id': 1, 'title': 'skip', 'count': 51, 'percent': 8.06}]"

In [16]:
def extract_rating_type(row):
#     json_data = ast.literal_eval(row['ratings'])
    rating_types=[]
    for item in row['ratings']:
        title = item['title']
        count = item['count']
        row[f'rating_type_{title}'] = count
        type_rating_cols.append(row[f'rating_type_{title}'])
    return row

df['ratings'] = df['ratings'].apply(convert_to_list)
df = df.apply(extract_rating_type, axis=1)

type_rating_cols=list(set(type_rating_cols))
df[type_rating_cols]=df[type_rating_cols].fillna(False)
df.dropt('ratings', inplace=True) # Cannot store a list of it, so, remove it
df.head(1)


added                                    added_by_status  \
0   6818  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...   

                                         esrb_rating genre_action  \
0  {'id': 3, 'name': 'Teen', 'slug': 'teen', 'nam...         True   

  genre_adventure genre_arcade genre_board-games genre_card genre_casual  \
0            True          NaN               NaN        NaN          NaN   

  genre_educational genre_family genre_fighting genre_indie  \
0               NaN          NaN            NaN         NaN   

  genre_massively-multiplayer genre_platformer genre_puzzle genre_racing  \
0                         NaN              NaN          NaN          NaN   

  genre_role-playing-games-rpg genre_shooter genre_simulation genre_sports  \
0                          NaN          True              NaN          NaN   

  genre_strategy                        genres     id  metacritic  \
0            NaN  [shooter, adventure, action]  58751        84.0   

            name                                   parent_platforms  \
0  Halo Infinite  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   

  platform_android platform_apple-ii platform_commodore-amiga  \
0              NaN               NaN                      NaN   

  platform_dreamcast platform_game-boy platform_game-boy-advance  \
0                NaN               NaN                       NaN   

  platform_game-boy-color platform_gamecube platform_genesis platform_ios  \
0                     NaN               NaN              NaN          NaN   

  platform_jaguar platform_linux platform_macintosh platform_macos  \
0             NaN            NaN                NaN            NaN   

  platform_neogeo platform_nes platform_nintendo-3ds platform_nintendo-64  \
0             NaN          NaN                   NaN                  NaN   

  platform_nintendo-ds platform_nintendo-dsi platform_nintendo-switch  \
0                  NaN                   NaN                      NaN   

  platform_pc platform_playstation1 platform_playstation2  \
0        True                   NaN                   NaN   

  platform_playstation3 platform_playstation4 platform_playstation5  \
0                   NaN                   NaN                   NaN   

  platform_ps-vita platform_psp platform_sega-cd platform_sega-master-system  \
0              NaN          NaN              NaN                         NaN   

  platform_sega-saturn platform_web platform_wii platform_wii-u  \
0                  NaN          NaN          NaN            NaN   

  platform_xbox-old platform_xbox-one platform_xbox-series-x platform_xbox360  \
0               NaN              True                   True              NaN   

                       platforms  playtime  rating  rating_top  \
0  [pc, xbox-one, xbox-series-x]         2    3.68           4   

   rating_type_exceptional  rating_type_meh  rating_type_recommended  \
0                    124.0            171.0                    287.0   

   rating_type_skip                                            ratings  \
0              51.0  [{'id': 4, 'title': 'recommended', 'count': 28...   

   ratings_count    released  reviews_count  score           slug  \
0            619  2021-12-08            633    NaN  halo-infinite   

  store_apple-appstore store_epic-games store_gog store_google-play  \
0                  NaN              NaN       NaN               NaN   

  store_itch store_nintendo store_playstation-store store_steam  \
0        NaN            NaN                     NaN        True   

  store_xbox-store store_xbox360               stores  user_game  
0             True           NaN  [steam, xbox-store]        NaN

## 5- ESRB (esrb_rating)
This code extract from ESRB the parental rating
 'slug': 'teen',
 As this is only one Rating, it could stay as str, and use a labelEncoder, instead of OneHotEncoding

In [17]:
rated_df['esrb_rating'].iloc[0]

"{'id': 3, 'name': 'Teen', 'slug': 'teen', 'name_en': 'Teen', 'name_ru': 'С 13 лет'}"

In [18]:
def convert_to_dict(column_data):
    try:
        return ast.literal_eval(column_data)
    except (ValueError, SyntaxError):
        return {}
    
def extract_esrb(row):
    esrb_dict = row['esrb_rating']
    row['esrb_rating'] = esrb_dict.get('name_en', '')  # Extract 'name_en' value from the dictionary
    return row
    

df['esrb_rating'] = df['esrb_rating'].astype(str)
df['esrb_rating'] = df['esrb_rating'].apply(convert_to_dict)
df = df.apply(extract_esrb, axis=1)
df['esrb_rating'].fillna("", inplace=True)
df.head(1)

added                                    added_by_status esrb_rating  \
0   6818  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...        Teen   

  genre_action genre_adventure genre_arcade genre_board-games genre_card  \
0         True            True          NaN               NaN        NaN   

  genre_casual genre_educational genre_family genre_fighting genre_indie  \
0          NaN               NaN          NaN            NaN         NaN   

  genre_massively-multiplayer genre_platformer genre_puzzle genre_racing  \
0                         NaN              NaN          NaN          NaN   

  genre_role-playing-games-rpg genre_shooter genre_simulation genre_sports  \
0                          NaN          True              NaN          NaN   

  genre_strategy                        genres     id  metacritic  \
0            NaN  [shooter, adventure, action]  58751        84.0   

            name                                   parent_platforms  \
0  Halo Infinite  [{'platform': {'id': 1, 'name': 'PC', 'slug': ...   

  platform_android platform_apple-ii platform_commodore-amiga  \
0              NaN               NaN                      NaN   

  platform_dreamcast platform_game-boy platform_game-boy-advance  \
0                NaN               NaN                       NaN   

  platform_game-boy-color platform_gamecube platform_genesis platform_ios  \
0                     NaN               NaN              NaN          NaN   

  platform_jaguar platform_linux platform_macintosh platform_macos  \
0             NaN            NaN                NaN            NaN   

  platform_neogeo platform_nes platform_nintendo-3ds platform_nintendo-64  \
0             NaN          NaN                   NaN                  NaN   

  platform_nintendo-ds platform_nintendo-dsi platform_nintendo-switch  \
0                  NaN                   NaN                      NaN   

  platform_pc platform_playstation1 platform_playstation2  \
0        True                   NaN                   NaN   

  platform_playstation3 platform_playstation4 platform_playstation5  \
0                   NaN                   NaN                   NaN   

  platform_ps-vita platform_psp platform_sega-cd platform_sega-master-system  \
0              NaN          NaN              NaN                         NaN   

  platform_sega-saturn platform_web platform_wii platform_wii-u  \
0                  NaN          NaN          NaN            NaN   

  platform_xbox-old platform_xbox-one platform_xbox-series-x platform_xbox360  \
0               NaN              True                   True              NaN   

                       platforms  playtime  rating  rating_top  \
0  [pc, xbox-one, xbox-series-x]         2    3.68           4   

   rating_type_exceptional  rating_type_meh  rating_type_recommended  \
0                    124.0            171.0                    287.0   

   rating_type_skip                                            ratings  \
0              51.0  [{'id': 4, 'title': 'recommended', 'count': 28...   

   ratings_count    released  reviews_count  score           slug  \
0            619  2021-12-08            633    NaN  halo-infinite   

  store_apple-appstore store_epic-games store_gog store_google-play  \
0                  NaN              NaN       NaN               NaN   

  store_itch store_nintendo store_playstation-store store_steam  \
0        NaN            NaN                     NaN        True   

  store_xbox-store store_xbox360               stores  user_game  
0             True           NaN  [steam, xbox-store]        NaN

        
### 6- Extract the parent Platform
Playstation 5 -> Playstation

In [19]:
rated_df['parent_platforms'].iloc[0]

"[{'platform': {'id': 1, 'name': 'PC', 'slug': 'pc'}}, {'platform': {'id': 3, 'name': 'Xbox', 'slug': 'xbox'}}]"

In [ ]:
parentplat_cols=[]
def extract_parent_platform(row):
    parent_plat_list = row['parent_platforms']
    slugs = [plat['platform']['slug'] for plat in parent_plat_list]
    row['parent_platforms'] = slugs
    for slug in slugs:
        row[f"parentplat_{slug}"] = True
        parentplat_cols.append(row[f"parentplat_{slug}"])
    return row

df['parent_platforms'] = df['parent_platforms'].apply(convert_to_list)
df = df.apply(extract_parent_platform, axis=1)
df['parent_platforms'].fillna("", inplace=True)

parentplat_cols=list(set(parentplat_cols))
df[parentplat_cols]=df[parentplat_cols].fillna(False)

df.head(1)

In [22]:
df.head(1)

added                                    added_by_status esrb_rating  \
0   6818  {'yet': 259, 'owned': 5280, 'beaten': 282, 'to...        Teen   

  genre_action genre_adventure genre_arcade genre_board-games genre_card  \
0         True            True          NaN               NaN        NaN   

  genre_casual genre_educational genre_family genre_fighting genre_indie  \
0          NaN               NaN          NaN            NaN         NaN   

  genre_massively-multiplayer genre_platformer genre_puzzle genre_racing  \
0                         NaN              NaN          NaN          NaN   

  genre_role-playing-games-rpg genre_shooter genre_simulation genre_sports  \
0                          NaN          True              NaN          NaN   

  genre_strategy                        genres     id  metacritic  \
0            NaN  [shooter, adventure, action]  58751        84.0   

            name parent_platforms parentplat_android parentplat_atari  \
0  Halo Infinite       [pc, xbox]                NaN              NaN   

  parentplat_commodore-amiga parentplat_ios parentplat_linux parentplat_mac  \
0                        NaN            NaN              NaN            NaN   

  parentplat_neo-geo parentplat_nintendo parentplat_pc parentplat_playstation  \
0                NaN                 NaN          True                    NaN   

  parentplat_sega parentplat_web parentplat_xbox platform_android  \
0             NaN            NaN            True              NaN   

  platform_apple-ii platform_commodore-amiga platform_dreamcast  \
0               NaN                      NaN                NaN   

  platform_game-boy platform_game-boy-advance platform_game-boy-color  \
0               NaN                       NaN                     NaN   

  platform_gamecube platform_genesis platform_ios platform_jaguar  \
0               NaN              NaN          NaN             NaN   

  platform_linux platform_macintosh platform_macos platform_neogeo  \
0            NaN                NaN            NaN             NaN   

  platform_nes platform_nintendo-3ds platform_nintendo-64  \
0          NaN                   NaN                  NaN   

  platform_nintendo-ds platform_nintendo-dsi platform_nintendo-switch  \
0                  NaN                   NaN                      NaN   

  platform_pc platform_playstation1 platform_playstation2  \
0        True                   NaN                   NaN   

  platform_playstation3 platform_playstation4 platform_playstation5  \
0                   NaN                   NaN                   NaN   

  platform_ps-vita platform_psp platform_sega-cd platform_sega-master-system  \
0              NaN          NaN              NaN                         NaN   

  platform_sega-saturn platform_web platform_wii platform_wii-u  \
0                  NaN          NaN          NaN            NaN   

  platform_xbox-old platform_xbox-one platform_xbox-series-x platform_xbox360  \
0               NaN              True                   True              NaN   

                       platforms  playtime  rating  rating_top  \
0  [pc, xbox-one, xbox-series-x]         2    3.68           4   

   rating_type_exceptional  rating_type_meh  rating_type_recommended  \
0                    124.0            171.0                    287.0   

   rating_type_skip                                            ratings  \
0              51.0  [{'id': 4, 'title': 'recommended', 'count': 28...   

   ratings_count    released  reviews_count  score           slug  \
0            619  2021-12-08            633    NaN  halo-infinite   

  store_apple-appstore store_epic-games store_gog store_google-play  \
0                  NaN              NaN       NaN               NaN   

  store_itch store_nintendo store_playstation-store store_steam  \
0        NaN            NaN                     NaN        True   

  store_xbox-store store_xbox360               stores  user_game  
0             True   


## By this point I should not have any column with a JSon or Dict format

## - Extract type of added_by_status
"{'yet': 259, 'owned': 5280, 'beaten': 282, 'toplay': 590, 'dropped': 295, 'playing': 112}"
how users have tagged this game in their library playlist, I will hel, but is not always complete for the amount of users
# Should we consider it, or not?